Rewrite systemd-swap in Python

oold opened the pull request:
The systemd-swap script was completely rewritten as a Python 3 script. Additionally, there are some minor changes to a few other files.

Notable changes:
- Script rewritten in Python 3.
- Makefile, unit file, man page, and package script updated accordingly.
- Updated date and version in man pages.
- Added `compression` command based on https://github.com/Nefelim4ag/systemd-swap/pull/93#issuecomment-577625769.

Closes #66 and closes #93.

Jarel1337 commented on systemd-swap.service:
```diff
@@ -1,9 +1,9 @@
 [Unit]
-Description=Manage swap spaces on zram, files and partitions.
+Description=Manage swap spaces on zram, files, and partitions.
```
Afaik there shouldn't be a comma here x2

oold replied:
It's called the [Oxford comma](https://www.grammarly.com/blog/what-is-the-oxford-comma-and-why-do-people-care-so-much-about-it/). It's entirely optional to use, but increases readability IMO. Anyway, I'm going to move those unrelated changes to different PRs.

Jarel1337 commented on package.sh:
```diff
@@ -22,11 +22,11 @@ debian_package(){
+    echo "Installed-Size: $(du -s "$DEB_NAME" | cut -f1)"
     echo "Maintainer: nefelim4ag@gmail.com"
-    echo "Description: Script for creating hybrid swap space from zram swaps, swap files and swap partitions."
+    echo "Description: Script for creating hybrid swap space from zram swaps, swap files, and swap partitions."
```
Afaik there shouldn't be a comma here

oold replied:
See https://github.com/Nefelim4ag/systemd-swap/pull/164#discussion_r456914847.

Jarel1337 commented on systemd-swap:
```diff
@@ -1,613 +1,724 @@
-
-exit 0
-# vim: set ft=sh ts=2 sw=2 et:
+#!/bin/env python3
```
Shouldn't this still be `#!/usr/bin/env`?

oold replied:
Correct.

Jarel1337 commented on systemd-swap:
```diff
@@ -1,613 +1,730 @@
-
-exit 0
-# vim: set ft=sh ts=2 sw=2 et:
+#!/usr/bin/env -S python3 -u
```
Actually, when looking at other Linux projects they all seem to use `#!/usr/bin/python{,3}` instead of `env`, I think we should follow that too

oold replied:
Looks like `#!/usr/bin/env -S python3 -u` is the preferred solution. `env` is always there and it will most likely find `python3`.

See:
- https://stackoverflow.com/a/19305076
- https://www.python.org/dev/peps/pep-0394/#for-python-script-publishers
- https://mail.python.org/pipermail/tutor/2007-June/054816.html

Jarel1337 replied:
The two python programs I have installed and running both use `/usr/bin/python{,3}` (tuned and firewalld).
The python.org pep linked only mentions using env for "scripts that are only expected to be run in an activated virtual environment" and since we're installing this as a system service this certainly isn't the case. 
![Htop screenshot](https://user-images.githubusercontent.com/26655508/87943046-c6128480-ca9d-11ea-93b6-2856a7a9a6e3.png)
Having `/usr/bin` also makes stuff like htop much more pleasing.
Since systemd-swap should be installed by the package manager and not pip it makes sense to use the system python installation which is always under `/usr/bin/python3`.

oold replied:
Updated the shebang.

oold commented:
Everything seems to be working fine, more or less. Only swapD needs testing now, but my dev system doesn't have a swap partition on my root disk and I currently don't want to set up a VM for this.

Jarel1337 commented on systemd-swap:
```diff
@@ -642,7 +642,8 @@ def stop(on_init: bool = False) -> None:
 
 def status() -> None:
-    init()
+    am_i_root()
```
This is totally unneeded, we don't actually need root for the status sub command (right?)

Jarel1337 replied:
Pff ignore me

Jarel1337 replied:
I do however think, from a design perspective or w/e that the status should be accessible to non root users aswell

oold replied:
Implemented. Only Zswap stats should be unavailable as non-root user.

Jarel1337 replied:
Oh thank you ^^, your doing all the work for me

oold replied:
You're welcome. It was only a small change and some testing.

Jarel1337 commented on systemd-swap.service:
```diff
@@ -16,6 +16,7 @@ DeviceAllow=block-blkext r
 DeviceAllow=block-device-mapper r
 DeviceAllow=block-loop rw
 DeviceAllow=block-sd r
+DeviceAllow=block-virtblk rw
```
Is write really necessary here? Only read permissions are needed for swap partitions

oold replied:
I'm going to test again.

oold replied:
No write access actually results in a `swapoff: Not superuser.`, so we seem to need write access.

Jarel1337 commented on man/systemd-swap.8:
```diff
@@ -1,9 +1,9 @@
-.TH systemd-swap 8 "JUNE 2020" "systemd-swap 4.2" systemd-swap
+.TH systemd-swap 8 "JULY 2020" "systemd-swap 5.0" systemd-swap
```
Actually, I was planning on having this as 4.4 following https://semver.org/
I'll do a write up of future plans and dreams in an issue later.

oold replied:
Oh, alright. I was going to ask, but assumed the changes might be too much to bump just the minor version. Especially the conf file handling is different now and might not always produce exactly the same results compared to the Bash version of systemd-swap.

Jarel1337 replied:
I agree that bumping to 5.0 would be valid, but I fear that we'd bump to 6.0 in under a year, and that feels like moving a bit too far. (Essentially the plan is to replace swapfc with [swapspace](https://github.com/Tookmund/Swapspace) for 5.0)

Jarel1337 replied:
We could implement both changes at the same time (have master be on python and only release 5.0 when swapspace is merged), but then the python rewrite gets delay a lot

oold replied:
Then I'd say adjusting only the minor version is fine. The changes are probably unlikely to break much unless some people got crazy with their `swap.conf` file.

Jarel1337 replied:
Great, I'll start looking and running the python code soon (no promises on time, but since it's such a large patch it'll probably take a week)

oold replied:
Thanks. I've changed the version to 4.4. It doesn't matter if it takes a week or longer, as long as we make sure that the rewrite didn't break anything.

oold commented:
> Just update the man pages to august +1and add a license header like the one [here](https://github.com/Nefelim4ag/systemd-swap/blob/rust_rewrite/src/main.rs#L1) (otherwise I'm assuming that it's GLPv3 or later)

Updated the man pages and added the license identifier. I also took the liberty to work on some code formatting. Black wants 88-character lines by default, but it doesn't seem to do a great job at enforcing that, so I manually reformatted the code according to that limit.

Jarel1337 commented:
Great, please squash my commit into one of yours and I'll merge it (don't squash anything else imo)

oold commented:
I'll just squash them all. No need to clutter the commit history.
